Initialize Language Server's JSON RPC protocol asynchronously

hubertp opened the pull request:
### Pull Request Description

As discovered in #7224, Json RPC protocol was added to the asynchronous resource initialization stage, as part of #6306, but was not in fact initialized at that point.
Instead it was initialized when the server was started to be able to serve correctly the initialization messages. A classic Catch-22. It was really hard to discover this just by looking at the code, but the profiling clearly showed where the time was spent.

This change splits Language Server's protocol into two:
- the first one accepts `heartbeat/init` and `session/initProtocolConnection`
- the second one enriches it with the full set of supported messages

This shifts the initialization from blocking for 0.5 sec to only ~30ms, and performing the second stage asynchronously.

Closes #7224.

### Important Notes

Before the change (blocking server startup):
![Screenshot from 2023-07-05 18-53-24](https://github.com/enso-org/enso/assets/292128/bcfa9043-d00a-4b36-a44c-782a388a16b9)
![Screenshot from 2023-07-05 18-53-10](https://github.com/enso-org/enso/assets/292128/54927787-4c95-46db-bd68-f3a3b82367d5)

After the change (1st stage):
![Screenshot from 2023-07-06 14-02-34](https://github.com/enso-org/enso/assets/292128/d7a7bc34-39dc-46f1-9e64-6d350697c30b)

After the change (2nd, asynchronous initialization, stage):
![Screenshot from 2023-07-06 14-21-17](https://github.com/enso-org/enso/assets/292128/def8c0a1-f211-4fc0-9df0-7c1634312166)


### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

hubertp committed:
Initialize Language Server's JSON RPC protocol asynchronously

As discovered in #7224, Json RPC protocol was added to the asynchronous
resource initialization stage, as part of #6306, but was not in fact
initialized at that point.
Instead it was initialized when the server was started to be able to
serve correctly the initialization messages. A classic Catch-22.
It was really hard to discover this just by looking at the code, but the
profiling clearly showed where the time was spent.

This change splits Language Server's protocol into two:
- the first one accepts `heartbeat/init` and
  `session/initProtocolConnection`
- the second one enriches it with the full set of supported messages

This shifts the initialization from blocking for 0.5 sec to only ~30ms,
and performing the second stage asynchronously.

4e6 commented on engine/language-server/src/main/scala/org/enso/languageserver/protocol/json/JsonRpc.scala:
```diff
@@ -29,12 +29,14 @@ import org.enso.languageserver.workspace.WorkspaceApi.ProjectInfo
+
+  /** A description of supported JSON RPC messages at a post-initialization stage */
+  def fullProtocol(init: Protocol): Protocol = init
+    .registerRequest(Ping)
```
The language server should be able to receive `InitProtocolConnection` after the initialization too https://github.com/enso-org/enso/blob/e7a0312298e212ee5c3ed58804eec44022515566/engine/language-server/src/main/scala/org/enso/languageserver/protocol/json/JsonConnectionController.scala#L283-L289

hubertp replied:
And it will, since it is already part of `init` argument.

4e6 commented on engine/language-server/src/main/scala/org/enso/languageserver/protocol/json/JsonRpcProtocolFactory.scala:
```diff
@@ -5,12 +5,21 @@ import org.enso.jsonrpc.{Protocol, ProtocolFactory}
 /** Factory creating JSON-RPC protocol. */
 final class JsonRpcProtocolFactory extends ProtocolFactory {
 
+  private[this] var _protocol: Protocol = _
```
`private` should be enough here, right? Just curious

hubertp replied:
It will, yes. A bit of a habit of mine to make variables object private by default.

hubertp committed:
nit

hubertp committed:
fix tests

Akirathan reviewed (approved):
Almost -500 ms from the startup. Nice.

hubertp committed:
Ensure correct error returned on missing method

When session has not been initialized we want to report a more
informative message.

JaroslavTulach reviewed (commented):
OK.
